src/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -99,19 +102,58 @@ public string GetHeaderValue(ReadOnlySpan<byte> headerValue)
             }
 
             // If it's a known header value, use the known value instead of allocating a new string.
-            if (_knownHeader != null && _knownHeader.KnownValues != null)
+            if (_knownHeader != null)
             {
-                string[] knownValues = _knownHeader.KnownValues;
-                for (int i = 0; i < knownValues.Length; i++)
+                if (_knownHeader.KnownValues != null)
+                {
+                    string[] knownValues = _knownHeader.KnownValues;
+                    for (int i = 0; i < knownValues.Length; i++)
+                    {
+                        if (ByteArrayHelpers.EqualsOrdinalAsciiIgnoreCase(knownValues[i], headerValue))
+                        {
+                            return knownValues[i];
+                        }
+                    }
+                }
+
+                if (KnownHeader == KnownHeaders.Location)
                 {
-                    if (ByteArrayHelpers.EqualsOrdinalAsciiIgnoreCase(knownValues[i], headerValue))
+                    // Normally Location should be in ISO-8859-1, ocassionally some servers respond with UTF-8 though
+                    if (TryDecodeUtf8(headerValue, out string decoded))
                     {
-                        return knownValues[i];
+                        return decoded;
                     }
                 }
             }
 
             return HttpRuleParser.DefaultHttpEncoding.GetString(headerValue);
         }
+
+        private static bool TryDecodeUtf8(ReadOnlySpan<byte> input, out string decoded)
+        {
+            // TODO: Utilize vectorization helpers here if/when they're made public
+            bool possibleUtf8 = false;
+            for (int i = 0; i < input.Length; i++)
+            {
+                if (input[i] > 127)
+                {
+                    possibleUtf8 = true;
+                    break;
+                }
+            }
+
+            if (possibleUtf8)
+            {
+                try
+                {
+                    decoded = s_utf8DecoderWithExceptionFallback.GetString(input);
+                    return true;
+                }
+                catch (ArgumentException) { } // Not actually Utf-8
```

Comment:
nit: "Utf-8" -> "UTF-8"

This makes the comments consistent since you used "UTF-8" in a comment above this.

Reply:
good catch!

Reply:
code path does not exist anymore, will leave this open until CI is green in case the API is unavailable somewhere